### PR TITLE
Returning dataframes

### PR DESCRIPTION
**Pull request by kantuni**
`pandas.DataFrame` can now be used as a widget.

### CODE REVIEW

File: examples/CustomDataframe/frontend/src/CustomDataframe.tsx [hunk 1]
```diff
@@ -18,10 +17,15 @@ interface TableProps {
   element: ArrowTable
 }
 
-class Table extends PureComponent<TableProps> {
+class Table extends React.PureComponent<TableProps> {
+  public componentDidMount = (): void => {
+    // Send the dataframe back to Python.
+    // NOTE: Styler will not be sent back as it's impossible to serialize.
+    Streamlit.setComponentValue(this.props.element.serialize(), "dataframe")
```
Comment (tconkling) [inline examples/CustomDataframe/frontend/src/CustomDataframe.tsx:24]: Is this just for illustrative purposes, or is there a reason you'd want to just return the dataframe back to Streamlit? (If it's just for illustration, I think we should not include it in this example - I think it'll be confusing to users. My first thought would be "is this something I need to do to make dataframes work?")

File: examples/CustomDataframe/frontend/src/streamlit/streamlit.ts [hunk 2]
```diff
@@ -115,10 +115,10 @@ export class Streamlit {
    * value = st.my_component(...)
    * st.write(value) # -> "ahoy!"
    *
-   * The value must be serializable into JSON.
+   * The value must be serializable into JSON or pandas.DataFrame.
    */
-  public static setComponentValue = (value: any): void => {
-    Streamlit.sendBackMsg(ComponentMessageType.SET_COMPONENT_VALUE, { value })
+  public static setComponentValue = (value: any, dataType = "json"): void => {
+    Streamlit.sendBackMsg(ComponentMessageType.SET_COMPONENT_VALUE, { value, dataType })
```
Comment (tconkling) [inline examples/CustomDataframe/frontend/src/streamlit/streamlit.ts:121]: Why is the `dataType` param required? Is it possible to introspect the value to see if its type is `ArrowTableProto` or `ArrowTable` instead?

File: examples/CustomDataframe/frontend/src/streamlit/streamlit.ts [hunk 3]
```diff
@@ -115,10 +115,10 @@ export class Streamlit {
    * value = st.my_component(...)
    * st.write(value) # -> "ahoy!"
    *
-   * The value must be serializable into JSON.
+   * The value must be serializable into JSON or pandas.DataFrame.
```
Comment (tconkling) [inline examples/CustomDataframe/frontend/src/streamlit/streamlit.ts:118]: pandas.DataFrame doesn't exist in javascript so this is a bit confusing. What about:

"The value must be serializable into JSON or ArrowTable" (or whatever version of that statement is most correct)

File: examples/CustomDataframe/frontend/src/CustomDataframe.tsx [hunk 4]
```diff
@@ -18,37 +12,46 @@ interface TableProps {
   element: ArrowTable
 }
 
-class Table extends PureComponent<TableProps> {
+class Table extends React.PureComponent<TableProps> {
+  public returnDataframe = (): void => {
+    // NOTE: Styler will not be sent back as it's impossible to serialize.
```
[the remark this answers is not in the document]
Comment (kantuni) [reply examples/CustomDataframe/frontend/src/CustomDataframe.tsx:17]: Done

**Pull request merged**